proof/crefine/RISCV64/StateRelation_C.thy

```diff
@@ -351,18 +363,78 @@ definition
+  "csched_context_relation asc csc \<equiv>
+       scPeriod asc = scPeriod_C csc
+     \<and> scConsumed asc = scConsumed_C csc
+     \<and> option_to_ptr (scTCB asc) = scTcb_C csc
+     \<and> option_to_ptr (scReply asc) = scReply_C csc
+     \<and> option_to_ptr (scNtfn asc) = scNotification_C csc
+     \<and> scBadge asc = scBadge_C csc
+     \<and> scSporadic asc = to_bool (scSporadic_C csc)
+     \<and> option_to_ptr (scYieldFrom asc) = scYieldFrom_C csc
+     \<and> scRefillMax asc = unat (scRefillMax_C csc)
+     \<and> scRefillHead asc = unat (scRefillHead_C csc)
+     \<and> scRefillCount asc = crefill_size csc"
+
+definition crefill_relation :: "refill \<Rightarrow> refill_C \<Rightarrow> bool" where
+  "crefill_relation ar cr \<equiv> rTime ar = rTime_C cr \<and> rAmount ar = rAmount_C cr"
+
+primrec dyn_array_list_rel ::
+  "('c :: c_type) typ_heap \<Rightarrow> ('a \<Rightarrow> 'c \<Rightarrow> bool) \<Rightarrow> 'a list \<Rightarrow> 'c ptr \<Rightarrow> bool"
+  where
+    "dyn_array_list_rel h rel [] p = True"
+  | "dyn_array_list_rel h rel (a # as) p
+      = (\<exists>v. h p = Some v \<and> rel a v \<and> dyn_array_list_rel h rel as (p +\<^sub>p 1))"
+
+definition sc_ptr_to_crefill_ptr :: "obj_ref \<Rightarrow> refill_C ptr" where
+  "sc_ptr_to_crefill_ptr p \<equiv> Ptr (p + of_nat sizeof_sched_context_t)"
+
+definition refill_buffer_relation ::
+  "(obj_ref \<rightharpoonup> kernel_object) \<Rightarrow> heap_raw_state \<Rightarrow> cghost_state \<Rightarrow> bool"
+  where
+  "refill_buffer_relation ah ch gs \<equiv>
+     let abs_sc_hp = map_to_scs ah;
+         crefill_hp = clift ch;
+         lens_hp = refill_buffer_lengths gs
+     in
+       dom abs_sc_hp = dom lens_hp
+     \<and> dom crefill_hp
+       = \<Union> ((\<lambda>p. set (map (\<lambda>i. sc_ptr_to_crefill_ptr p +\<^sub>p int i) [0..< the (lens_hp p)])) ` (dom abs_sc_hp))
+     \<and> (\<forall>p sc. abs_sc_hp p = Some sc \<longrightarrow>
+                 dyn_array_list_rel crefill_hp crefill_relation (scRefills sc) (sc_ptr_to_crefill_ptr p)
+                 \<and> lens_hp p = Some (length (scRefills sc)))"
+
+definition creply_relation :: "Structures_H.reply \<Rightarrow> reply_C \<Rightarrow> bool" where
+  "creply_relation areply creply \<equiv>
+       option_to_ptr (replyTCB areply) = replyTCB_C creply
+     \<and> option_to_ptr (replyPrev areply)
+        = ((Ptr \<circ> callStackPtr_CL \<circ> call_stack_lift \<circ> replyPrev_C) creply :: reply_C ptr)
+     \<and> isHead (replyNext areply) = to_bool ((isHead_CL \<circ> call_stack_lift \<circ> replyNext_C) creply)
+     \<and> (if bound (replyNext areply)
+        then if isHead (replyNext areply)
+             then theHeadScPtr (replyNext areply)
+             else theReplyNextPtr (replyNext areply)
+        else 0)
+       = (callStackPtr_CL \<circ> call_stack_lift \<circ> replyNext_C) creply"
```

Comment:
Stupid question: what is `call_stack_lift`? I can't find a definition.

Reply:
I can't open up CRefine just yet to double check and point you to the exact location, but `call_stack` is a type defined via a C `block` statement and it's in the same file as the others, like `endpoint_lift`. Can you find it now?

Reply:
Right, thanks for that, and of course that makes perfect sense. All good.